packages/runner/src/storage/transaction/chronicle.ts

```diff
@@ -541,68 +526,154 @@ class Novelty {
-        }
+        // Return the working copy which has the merged state
+        return this.#workingCopy;
       }
     }
+    return undefined;
+  }
 
-    return candidate;
+  /**
+   * Apply a write directly to the working copy - O(1) amortized.
+   * This replaces the old put() + rebase() pattern that was O(N).
+   */
+  applyWrite(
+    address: IMemoryAddress,
+    value: JSONValue | undefined,
+  ): Result<
+    IAttestation,
+    IStorageTransactionInconsistent | INotFoundError | ITypeMismatchError
+  > {
+    if (!this.#workingCopy) {
+      return {
+        error: {
+          name: "StorageTransactionInconsistent",
+          message: "Cannot apply write without initialized working copy",
+          address,
+          from: () => ({
+            name: "StorageTransactionInconsistent",
+            message: "Cannot apply write without initialized working copy",
+            address,
+            from: function () {
+              return this;
+            },
+          }),
+        },
+      };
+    }
+
+    const result = write(this.#workingCopy, address, value);
+    if (result.ok) {
+      this.#workingCopy = result.ok;
+      // Store individual path attestation for novelty() iterator
+      const pathKey = JSON.stringify(address.path);
+      this.#pathAttestations.set(pathKey, { address, value });
+    }
+    return result;
   }
 
+  /** Legacy put() for compatibility - applies write to working copy */
   put(invariant: IAttestation) {
-    this.#model.set(JSON.stringify(invariant.address.path), invariant);
+    if (!this.#workingCopy) {
+      // First write initializes the working copy
+      this.#workingCopy = invariant;
+    } else {
+      // Apply write to working copy
+      const result = write(
+        this.#workingCopy,
+        invariant.address,
+        invariant.value,
+      );
+      if (result.ok) {
+        this.#workingCopy = result.ok;
+      }
+    }
+    // Store individual path attestation for novelty() iterator
+    const pathKey = JSON.stringify(invariant.address.path);
+    this.#pathAttestations.set(pathKey, invariant);
   }
+
+  /** Legacy delete() - removes from path attestations */
   delete(invariant: IAttestation) {
-    this.#model.delete(JSON.stringify(invariant.address.path));
+    const pathKey = JSON.stringify(invariant.address.path);
+    this.#pathAttestations.delete(pathKey);
   }
 
   /**
-   * Applies all the overlapping write invariants onto a given source invariant.
+   * Get the working copy. Used for commit and reads.
    */
+  getWorkingCopy(): IAttestation | undefined {
+    return this.#workingCopy;
+  }
 
+  /**
+   * Returns the working copy as the rebased result.
+   * With the working copy pattern, no replay is needed - O(1).
+   *
+   * CT-1123: This was the O(N²) hotspot - each call iterated all changes
+   * and deep-cloned. Now it just returns the already-merged state.
+   */
   rebase(
-    source: IAttestation,
+    _source: IAttestation,
```

Comment:
leave it as `source`, since we're actually using it below in the return statement.